src/Raven.Server/Documents/Versioning/VersioningStorage.cs

```diff
@@ -0,0 +1,257 @@
+            BlittableJsonReaderObject document, BlittableJsonReaderObject oldDocument, bool isSystemDocument)
+        {
+            if (isSystemDocument)
+                return;
+
+            bool enableVersioning = false;
+            BlittableJsonReaderObject metadata;
+            if (document != null && document.TryGet(Constants.Metadata, out metadata))
+            {
+                if (metadata.TryGet(Constants.Versioning.RavenEnableVersioning, out enableVersioning))
+                {
+                    metadata.Modifications.Remove(Constants.Versioning.RavenEnableVersioning);
+                }
+
+                bool disableVersioning;
+                if (metadata.TryGet(Constants.Versioning.RavenDisableVersioning, out disableVersioning))
+                {
+                    metadata.Modifications.Remove(Constants.Versioning.RavenDisableVersioning);
+                    if (disableVersioning)
+                        return;
+                }
+            }
+
+            BlittableJsonReaderObject configuration;
+            if (IsVersioningActive(collectionName, enableVersioning, out configuration) == false)
+                return;
+
+            var table = new Table(_docsSchema, "_revisions/" + collectionName, context.Transaction.InnerTransaction);
+
+            int? maxRevisions;
+            configuration.TryGet("MaxRevisions", out maxRevisions);
+            var revisionsCount = IncrementCountOfRevisions(context, key, 1);
+            DeleteOldRevisions(context, table, collectionName, key, maxRevisions, revisionsCount);
+
+            byte* lowerKey;
+            int lowerSize;
+            byte* keyPtr;
+            int keySize;
+            DocumentsStorage.GetLowerKeySliceAndStorageKey(context, key, out lowerKey, out lowerSize, out keyPtr, out keySize);
+
+            var newEtag = ++_lastEtag;
+            var newEtagBigEndian = IPAddress.HostToNetworkOrder(newEtag);
+
+            var tbv = new TableValueBuilder
+            {
+                {lowerKey, lowerSize},
+                {(byte*)&newEtagBigEndian, sizeof(long)},
+                {keyPtr, keySize},
+                {oldDocument.BasePointer, oldDocument.Size}
+            };
+
+            var insert = table.Insert(tbv);
+        }
+
+        private void DeleteOldRevisions(DocumentsOperationContext context, Table table, string collectionName, string key, int? maxRevisions, long revisionsCount)
+        {
+            if (maxRevisions.HasValue == false || maxRevisions.Value == int.MaxValue)
+                return;
+
+            var numberOfRevisionsToDelete = revisionsCount - maxRevisions.Value;
+            if (numberOfRevisionsToDelete <= 0)
+                return;
+
+            var deletedRevisionsCount = table.DeleteForwardFrom(_docsSchema.Indexes["KeyAndEtag"], key, numberOfRevisionsToDelete);
+            IncrementCountOfRevisions(context, collectionName, key, -deletedRevisionsCount);
```

Comment:
This is a good place for Debug.Assert because numberOfRevisionsToDelete and deletedRevisionsCount must be equal. Or I'm wrong?